include/grpc++/impl/codegen/thrift_serializer.h

```diff
@@ -0,0 +1,223 @@
+#include <grpc/impl/codegen/slice.h>
+#include <grpc/impl/codegen/slice_buffer.h>
+#include <thrift/protocol/TBinaryProtocol.h>
+#include <thrift/protocol/TCompactProtocol.h>
+#include <thrift/protocol/TProtocolException.h>
+#include <thrift/transport/TBufferTransports.h>
+#include <thrift/transport/TTransportUtils.h>
+#include <boost/make_shared.hpp>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace apache {
+namespace thrift {
+namespace util {
+
+using apache::thrift::protocol::TBinaryProtocolT;
+using apache::thrift::protocol::TCompactProtocolT;
+using apache::thrift::protocol::TMessageType;
+using apache::thrift::protocol::TNetworkBigEndian;
+using apache::thrift::transport::TMemoryBuffer;
+using apache::thrift::transport::TBufferBase;
+using apache::thrift::transport::TTransport;
+
+template <typename Dummy, typename Protocol>
+class ThriftSerializer {
+ public:
+  ThriftSerializer()
+      : prepared_(false),
+        last_deserialized_(false),
+        serialize_version_(false) {}
+
+  virtual ~ThriftSerializer() {}
+
+  // Serialize the passed type into the internal buffer
+  // and returns a pointer to internal buffer and its size
+  template <typename T>
+  void Serialize(const T& fields, const uint8_t** serialized_buffer,
+                 size_t* serialized_len) {
+    // prepare or reset buffer
+    if (!prepared_ || last_deserialized_) {
+      prepare();
+    } else {
+      buffer_->resetBuffer();
+    }
+    last_deserialized_ = false;
+
+    // if required serialize protocol version
+    if (serialize_version_) {
+      protocol_->writeMessageBegin("", TMessageType(0), 0);
+    }
+
+    // serialize fields into buffer
+    fields.write(protocol_.get());
+
+    // write the end of message
+    if (serialize_version_) {
+      protocol_->writeMessageEnd();
+    }
+
+    uint8_t* byte_buffer;
+    uint32_t byte_buffer_size;
+    buffer_->getBuffer(&byte_buffer, &byte_buffer_size);
+    *serialized_buffer = byte_buffer;
+    *serialized_len = byte_buffer_size;
+  }
+
+  // Serialize the passed type into the byte buffer
+  template <typename T>
+  void Serialize(const T& fields, grpc_byte_buffer** bp) {
+    const uint8_t* byte_buffer;
+    size_t byte_buffer_size;
+
+    Serialize(fields, &byte_buffer, &byte_buffer_size);
+
+    gpr_slice slice =
+        gpr_slice_from_copied_buffer((char*)byte_buffer, byte_buffer_size);
+
+    *bp = grpc_raw_byte_buffer_create(&slice, 1);
+
+    gpr_slice_unref(slice);
+  }
+
+  // Deserialize the passed char array into  the passed type, returns the number
+  // of bytes that have been consumed from the passed string.
+  template <typename T>
+  uint32_t Deserialize(const uint8_t* serialized_buffer, size_t length,
+                       T* fields) {
+    // prepare buffer if necessary
+    if (!prepared_) {
+      prepare();
+    }
+    last_deserialized_ = true;
+
+    // reset buffer transport
+    buffer_->resetBuffer((uint8_t*)serialized_buffer, length);
```

Comment:
It's const pointer . We are not changing the memory location but the value
at that memory location.

On Thursday, August 4, 2016, Vijay Pai <notifications@github.com> wrote:

> In include/grpc++/impl/codegen/thrift_serializer.h
> <https://github.com/grpc/grpc/pull/7594#discussion_r73531176>:
>
> > +    gpr_slice_unref(slice);
> > +  }
> > +
> > +  // Deserialize the passed char array into  the passed type, returns the number
> > +  // of bytes that have been consumed from the passed string.
> > +  template <typename T>
> > +  uint32_t Deserialize(const uint8_t* serialized_buffer, size_t length,
> > +                       T* fields) {
> > +    // prepare buffer if necessary
> > +    if (!prepared_) {
> > +      prepare();
> > +    }
> > +    last_deserialized_ = true;
> > +
> > +    // reset buffer transport
> > +    buffer_->resetBuffer((uint8_t*)serialized_buffer, length);
>
> It's not that. In this case, you're trying to convert from a const to a
> non-const. So you should be using a const-cast to strip the const-ness. const_cast<uint8_t
> *> . That said, why is this situation arising? Why does Deserialize have
> a const argument for something that will actually get changed?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/7594/files/2e698457f370091111d0fe620bad306c743a3548#r73531176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASfuIgQefAryTLjxPpDePDSNk7zMMEHPks5qcfWugaJpZM4JZPv_>
> .
>